File: shared/data/src/debrief_data/registry.py
"""Platform registry loader — reads the vessel class tree and resolves platforms."""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

_REGISTRY_PATH = Path(__file__).resolve().parent.parent.parent / "platform-registry.json"


@dataclass(frozen=True)
class ResolvedPlatform:
    """Complete metadata for a platform, combining leaf attributes with position-derived fields."""

    id: str
    name: str
    nationality: str
    vessel_class: str
    vessel_type: str
    vessel_role: str
    domain: str
    short_name: str | None = None


class RegistryError(ValueError):
    """Raised when the registry file is structurally invalid."""


class PlatformRegistry:
    """Parsed platform registry with lookup, enumeration, and tree traversal."""

    def __init__(self, platforms: dict[str, ResolvedPlatform], tree: dict[str, Any]) -> None:
        self._platforms = platforms
        self._tree = tree

    def resolve(self, platform_id: str) -> ResolvedPlatform | None:
        """Look up a single platform by ID. Returns None for unknown/empty IDs."""
        if not platform_id or not platform_id.strip():
            return None
        return self._platforms.get(platform_id)

    def list_platforms(self) -> list[ResolvedPlatform]:
        """Return all registered platforms sorted by ID."""
        return sorted(self._platforms.values(), key=lambda p: p.id)

    def find_by_class(self, class_path: str) -> list[ResolvedPlatform]:
        """Find all platforms under a given vessel class path (including descendants)."""
        if not class_path or not class_path.strip():
            return []
        segments = class_path.strip().split("/")
        node = self._tree
        for seg in segments:
            if seg not in node:
                return []
            child = node[seg]
            if not isinstance(child, dict):
                return []
            node = child
        results: list[ResolvedPlatform] = []
        _collect_platforms(node, results, self._platforms)
        return sorted(results, key=lambda p: p.id)

    def is_valid_class(self, class_path: str) -> bool:
        """Check whether a class path corresponds to a real node in the taxonomy tree."""
        if not class_path or not class_path.strip():
            return False
        segments = class_path.strip().split("/")
        node = self._tree
        for seg in segments:
            if seg not in node:
                return False
            child = node[seg]
            if not isinstance(child, dict):
                return False
            node = child
        return True


def _collect_platforms(
    node: dict[str, Any],
    results: list[ResolvedPlatform],
    index: dict[str, ResolvedPlatform],
) -> None:
    """Recursively collect all platforms at or below a tree node."""
    for key, value in node.items():
        if key.startswith("_"):
            continue
        if not isinstance(value, dict):
            continue
        if "name" in value:
            platform = index.get(key)
            if platform is not None:
                results.append(platform)
        else:
            _collect_platforms(value, results, index)


def _is_platform_entry(value: Any) -> bool:
    """A node is a platform if it's a dict with a 'name' field."""
    return isinstance(value, dict) and "name" in value


def _is_leaf_entry(value: Any) -> bool:
    """A node is a leaf if it's a dict where any non-underscore value is not a dict.

    This catches malformed platforms (e.g., missing 'name' but has 'nationality').
    """
    if not isinstance(value, dict):
        return False
    for k, v in value.items():
        if k.startswith("_"):
            continue
        if not isinstance(v, dict):
            return True
    return False


def _walk_tree(
    node: dict[str, Any],
    path_segments: list[str],
    platforms: dict[str, ResolvedPlatform],
    seen_ids: dict[str, str],
) -> None:
    """Walk the vessel class tree, extracting platforms and building the index."""
    for key, value in node.items():
        if key.startswith("_"):
            continue
        if not isinstance(value, dict):
            raise RegistryError(
                f"Invalid registry format: expected object for key '{key}' "
                f"at path '{'/'.join(path_segments)}'"
            )
        if _is_platform_entry(value) or _is_leaf_entry(value):
            # Validate required fields
            if "name" not in value or not value["name"]:
                raise RegistryError(f"Platform '{key}' missing required field 'name'")
            if "nationality" not in value or not value["nationality"]:
                raise RegistryError(f"Platform '{key}' missing required field 'nationality'")

            # Check for duplicate IDs
            current_path = "/".join(path_segments)
            if key in seen_ids:
                raise RegistryError(
                    f"Duplicate platform ID '{key}' found at paths "
                    f"'{seen_ids[key]}' and '{current_path}'"
                )
            seen_ids[key] = current_path

            # Derive positional fields
            class_path = "/".join(path_segments)
            domain = path_segments[0] if len(path_segments) >= 1 else ""
            vessel_type = path_segments[-1] if len(path_segments) >= 1 else ""
            vessel_role = path_segments[-2] if len(path_segments) >= 2 else ""

            platforms[key] = ResolvedPlatform(
                id=key,
                name=value["name"],
                short_name=value.get("short_name"),
                nationality=value["nationality"],
                vessel_class=class_path,
                vessel_type=vessel_type,
                vessel_role=vessel_role,
                domain=domain,
            )
        else:
            _walk_tree(value, [*path_segments, key], platforms, seen_ids)


def load_registry(path: str | Path | None = None) -> PlatformRegistry:
    """Load and validate the platform registry from a JSON file.

    Args:
        path: Path to the registry JSON file. Defaults to the bundled registry.

    Returns:
        A PlatformRegistry instance with all platforms indexed.

    Raises:
        FileNotFoundError: If the registry file does not exist.
        RegistryError: If the file content is invalid.
    """
    registry_path = Path(path) if path is not None else _REGISTRY_PATH

    if not registry_path.exists():
        raise FileNotFoundError(f"Registry file not found: {registry_path}")

    text = registry_path.read_text(encoding="utf-8")
    try:
        data = json.loads(text)
    except json.JSONDecodeError as exc:
        raise RegistryError(f"Invalid registry format: {exc}") from exc

    if not isinstance(data, dict) or "vessel_classes" not in data:
        raise RegistryError("Registry must have 'vessel_classes' root key")

    vessel_classes = data["vessel_classes"]
    if not isinstance(vessel_classes, dict):
        raise RegistryError("Invalid registry format: 'vessel_classes' must be an object")

    platforms: dict[str, ResolvedPlatform] = {}
    seen_ids: dict[str, str] = {}
    _walk_tree(vessel_classes, [], platforms, seen_ids)

    return PlatformRegistry(platforms, vessel_classes)

File: shared/data/src/ts/registry.ts
/**
 * Platform registry loader — reads the vessel class tree and resolves platforms.
 */

import { readFileSync } from 'node:fs';
import { resolve as resolvePath, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const DEFAULT_REGISTRY_PATH = resolvePath(__dirname, '..', '..', 'platform-registry.json');

/** Complete metadata for a platform, combining leaf attributes with position-derived fields. */
export interface ResolvedPlatform {
  readonly id: string;
  readonly name: string;
  readonly short_name: string | undefined;
  readonly nationality: string;
  readonly vessel_class: string;
  readonly vessel_type: string;
  readonly vessel_role: string;
  readonly domain: string;
}

type TreeNode = Record<string, unknown>;

function isPlatformEntry(value: unknown): value is { name: string; nationality: string; short_name?: string } {
  return typeof value === 'object' && value !== null && 'name' in value;
}

function isLeafEntry(value: Record<string, unknown>): boolean {
  for (const [k, v] of Object.entries(value)) {
    if (k.startsWith('_')) continue;
    if (typeof v !== 'object' || v === null) return true;
  }
  return false;
}

function walkTree(
  node: TreeNode,
  pathSegments: readonly string[],
  platforms: Map<string, ResolvedPlatform>,
  seenIds: Map<string, string>,
): void {
  for (const [key, value] of Object.entries(node)) {
    if (key.startsWith('_')) continue;

    if (typeof value !== 'object' || value === null) {
      throw new Error(
        `Invalid registry format: expected object for key '${key}' at path '${pathSegments.join('/')}'`,
      );
    }

    const record = value as Record<string, unknown>;

    if (isPlatformEntry(record) || isLeafEntry(record)) {
      // Validate required fields
      if (!record.name || typeof record.name !== 'string') {
        throw new Error(`Platform '${key}' missing required field 'name'`);
      }
      if (!record.nationality || typeof record.nationality !== 'string') {
        throw new Error(`Platform '${key}' missing required field 'nationality'`);
      }

      // Check for duplicate IDs
      const currentPath = pathSegments.join('/');
      const existingPath = seenIds.get(key);
      if (existingPath !== undefined) {
        throw new Error(
          `Duplicate platform ID '${key}' found at paths '${existingPath}' and '${currentPath}'`,
        );
      }
      seenIds.set(key, currentPath);

      // Derive positional fields
      const classPath = pathSegments.join('/');
      const domain = pathSegments[0] ?? '';
      const vesselType = pathSegments[pathSegments.length - 1] ?? '';
      const vesselRole = pathSegments.length >= 2 ? (pathSegments[pathSegments.length - 2] ?? '') : '';

      platforms.set(key, {
        id: key,
        name: record.name as string,
        short_name: typeof record.short_name === 'string' ? record.short_name : undefined,
        nationality: record.nationality as string,
        vessel_class: classPath,
        vessel_type: vesselType,
        vessel_role: vesselRole,
        domain,
      });
    } else {
      walkTree(record, [...pathSegments, key], platforms, seenIds);
    }
  }
}

function collectPlatforms(
  node: TreeNode,
  results: ResolvedPlatform[],
  index: Map<string, ResolvedPlatform>,
): void {
  for (const [key, value] of Object.entries(node)) {
    if (key.startsWith('_')) continue;
    if (typeof value !== 'object' || value === null) continue;

    const record = value as Record<string, unknown>;
    if (isPlatformEntry(record)) {
      const platform = index.get(key);
      if (platform !== undefined) {
        results.push(platform);
      }
    } else {
      collectPlatforms(record, results, index);
    }
  }
}

/** Parsed platform registry with lookup, enumeration, and tree traversal. */
export class PlatformRegistry {
  private readonly _platforms: Map<string, ResolvedPlatform>;
  private readonly _tree: TreeNode;

  constructor(platforms: Map<string, ResolvedPlatform>, tree: TreeNode) {
    this._platforms = platforms;
    this._tree = tree;
  }

  /** Look up a single platform by ID. Returns undefined for unknown/empty IDs. */
  resolve(platformId: string): ResolvedPlatform | undefined {
    if (!platformId || !platformId.trim()) return undefined;
    return this._platforms.get(platformId);
  }

  /** Return all registered platforms sorted by ID. */
  listPlatforms(): readonly ResolvedPlatform[] {
    return [...this._platforms.values()].sort((a, b) => a.id.localeCompare(b.id));
  }

  /** Find all platforms under a given vessel class path (including descendants). */
  findByClass(classPath: string): readonly ResolvedPlatform[] {
    if (!classPath || !classPath.trim()) return [];
    const segments = classPath.trim().split('/');
    let node: TreeNode = this._tree;
    for (const seg of segments) {
      const child = node[seg];
      if (child === undefined || typeof child !== 'object' || child === null) return [];
      node = child as TreeNode;
    }
    const results: ResolvedPlatform[] = [];
    collectPlatforms(node, results, this._platforms);
    return results.sort((a, b) => a.id.localeCompare(b.id));
  }

  /** Check whether a class path corresponds to a real node in the taxonomy tree. */
  isValidClass(classPath: string): boolean {
    if (!classPath || !classPath.trim()) return false;
    const segments = classPath.trim().split('/');
    let node: TreeNode = this._tree;
    for (const seg of segments) {
      const child = node[seg];
      if (child === undefined || typeof child !== 'object' || child === null) return false;
      node = child as TreeNode;
    }
    return true;
  }
}

/**
 * Load and validate the platform registry from a JSON file.
 *
 * @param path - Path to the registry JSON file. Defaults to the bundled registry.
 * @returns A PlatformRegistry instance with all platforms indexed.
 * @throws Error if the file does not exist, contains invalid JSON, or has structural issues.
 */
export function loadRegistry(path?: string): PlatformRegistry {
  const registryPath = path ?? DEFAULT_REGISTRY_PATH;

  let text: string;
  try {
    text = readFileSync(registryPath, 'utf-8');
  } catch (err) {
    if (err instanceof Error && 'code' in err && (err as NodeJS.ErrnoException).code === 'ENOENT') {
      throw new Error(`Registry file not found: ${registryPath}`);
    }
    throw err;
  }

  let data: unknown;
  try {
    data = JSON.parse(text);
  } catch {
    throw new Error('Invalid registry format: invalid JSON');
  }

  if (typeof data !== 'object' || data === null || !('vessel_classes' in data)) {
    throw new Error("Registry must have 'vessel_classes' root key");
  }

  const vesselClasses = (data as Record<string, unknown>).vessel_classes;
  if (typeof vesselClasses !== 'object' || vesselClasses === null) {
    throw new Error("Invalid registry format: 'vessel_classes' must be an object");
  }

  const platforms = new Map<string, ResolvedPlatform>();
  const seenIds = new Map<string, string>();
  walkTree(vesselClasses as TreeNode, [], platforms, seenIds);

  return new PlatformRegistry(platforms, vesselClasses as TreeNode);
}
